prereise/gather/griddata/hifld/data_process/generators.py

```diff
@@ -22,54 +24,108 @@ def floatify(value, default=float("nan")):
+    """Determine the closest substation to each generator. For generators without
+    latitude and longitude, an attempt will be made to match via ZIP code, and failing
+    that a pandas.NA value will be returned.
+
+    :param pandas.DataFrame generators: generator data. Required columns:
+        'interconnect', 'lat', 'lon', 'state', 'ZIP'.
+    :param pandas.DataFrame substations: substation data. Required columns:
+        'interconnect', 'lat', 'lon', 'STATE', 'ZIP'.
+    :param bool inplace: whether to modify the generator table inplace with a new
+        'sub_id' column or to return a new one. If ``inplace`` is `True`, entries in
+        `generators` which have non-sensical combinations of 'state' and 'interconnect'
+        columns will have their 'interconnect' entries modified.
+    :return: (*pandas.DataFrame/None*) -- if ``inplace`` is `False`, return the modified
+        DataFrame; otherwise return nothing.
     """
-    lookup_params = tuple(generator.loc[["interconnect", "ZIP"]])
-    if pd.isna(generator["lat"]) or pd.isna(generator["lon"]):
-        # No location available
-        try:
-            matching_subs = substation_groupby.get_group(lookup_params)
-            return matching_subs.index[0]
-        except KeyError:
-            return pd.NA
-    try:
-        # This ZIP code contains substations, this block will execute successfully
-        matching_subs = substation_groupby.get_group(lookup_params)
-    except KeyError:
-        # If this ZIP code does not contain substations, this block will execute, and
-        # we select a set of 'nearby' substations
-        zip_range = [int(generator.loc["ZIP"]) + offset for offset in range(-100, 101)]
-        zip_range_strings = [str(z).rjust(5, "0") for z in zip_range]
-        try:
-            matching_subs = pd.concat(
-                [
-                    substation_groupby.get_group((generator.loc["interconnect"], z))
-                    for z in zip_range_strings
-                    if (generator.loc["interconnect"], z) in substation_groupby.groups
-                ]
-            )
-        except ValueError:
-            # If no matching subs within the given interconnection and ZIPs, give up
+
+    def get_sub_id_of_closest_substation(generator, state_trees, subs_state_lookup):
+        if not isinstance(generator["xyz"], list):
             return pd.NA
-    distance_to_subs = matching_subs.apply(
-        lambda x: haversine((x.lat, x.lon), (generator.lat, generator.lon)),
+        grouper_key = (generator["interconnect"], generator["state"])
+        _, array_index = state_trees[grouper_key].query(generator["xyz"])
+        sub_index = subs_state_lookup[grouper_key][array_index]
+        return sub_index
+
+    # Translate lat/lon to 3D positions (assume spherical earth, origin at center)
+    substations_with_xyz = substations.assign(
+        xyz=substations.apply(lambda x: latlon_to_xyz(x["lat"], x["lon"]), axis=1)
+    )
```

Comment:
I think we can simply do 
`substations_with_xyz = substations.assign(xyz=lambda x: latlon_to_xyz(x.lat, x.lon))` here.

Reply:
Never mind...I looked into this and found it doesn't work given `assign` + `lambda` work with columns only. Let's keep the current implementation.